File: packages/audio/android/src/main/java/dev/tcode/thinmpr/audio/service/ArtistService.kt
package dev.tcode.thinmpr.audio.service

import android.content.Context
import dev.tcode.thinmpr.audio.extension.toMap
import dev.tcode.thinmpr.audio.model.valueObject.ArtistId
import dev.tcode.thinmpr.audio.model.valueObject.SongId
import dev.tcode.thinmpr.audio.repository.AlbumRepository
import dev.tcode.thinmpr.audio.repository.ArtistRepository
import dev.tcode.thinmpr.audio.service.contract.ArtistServiceContract

class ArtistService(context: Context): ArtistServiceContract {
    private val artistRepository = ArtistRepository(context)
    private val albumRepository = AlbumRepository(context)

    override fun getAllArtists(): List<Map<String, Any>> {
        val artists = artistRepository.findAll()

        return artists.map { it.toMap() }
    }

    override fun getArtistDetailById(id: String): Map<String, Any>? {
        val artistId = ArtistId(id)
        val artist = artistRepository.findById(artistId) ?: return null
        val album = albumRepository.findFirstByArtistId(artistId)
        val name = artist?.name ?: ""
        val imageId = album?.imageId ?: ""

        return mapOf(
            "id" to artistId.raw,
            "name" to name,
            "imageId" to imageId,
        )
    }

    override fun getArtistsByIds(ids: List<String>): List<Map<String, Any>> {
        val artistIds = ids.map { ArtistId(it) }
        val artists = artistRepository.findByIds(artistIds)

        return artists.map { it.toMap() }
    }
}

File: packages/audio/ios/Service/ArtistService.swift
class ArtistService: ArtistServiceContract {
    private let artistRepository = ArtistRepository()

    func getAllArtists() -> [[String: Any]] {
        let artists = artistRepository.findAll()

        return artists.map { $0.toMap() }
    }

    func getArtistDetailById(id: String) -> [String: Any]? {
        let artistId = ArtistId(id: id)
        let albumRepository = AlbumRepository()
        let artist = artistRepository.findById(artistId: artistId)

        if artist == nil {
            return nil
        }

        let album = albumRepository.findFirstByArtistId(artistId: artistId)
        let name = artist?.name ?? ""
        let imageId = album?.imageId ?? ""

        return [
            "id": id,
            "name": name,
            "imageId": imageId,
        ]
    }

    func getArtistsByIds(ids: [String]) -> [[String: Any]] {
        let artistIds = ids.map { ArtistId(id: $0) }
        let artists = artistRepository.findByIds(artistIds: artistIds)

        return artists.map { $0.toMap() }
    }
}
